controllers/runnerdeployment_controller.go

```diff
@@ -270,6 +316,12 @@ func (r *RunnerDeploymentReconciler) newRunnerReplicaSet(rd v1alpha1.RunnerDeplo
 
 	newRSTemplate.Labels = labels
 
+	if rd.Spec.Selector == nil {
```

Comment:
This seems like a breaking change and I'd like to make it backward-compatible as much as possible.

Sorry if I'm missing something or I have any misconception - but can we use `labels` to compose the default rd.Spec.Selector here? As far as remember, `labels` is inherited to RunnerReplicaSet.ObjectMeta.Labels and and Runner.ObjectMeta.Labels, which might be enough for resolving the original issue.

As far as it is backward compatible. I think I'm fine to add `selector` and `matchLabels`, `matchExpressions`. IMHO it isn't strictly needed for the original issue, though. WDYT?

Reply:
I've tried to express the above as code in 62fcd5f. Does it make sense to you?

Reply:
[62fcd5f](https://github.com/summerwind/actions-runner-controller/commit/62fcd5f3de285f2c1384406fd0efbbfad3f01a92) basically looks good, but it's not completely backward-compatiblem, IMO.

Let's say users upgrades the controller while `RunnerReplicaSet`s already exist.
The new controller cannot find the existing `RunnerReplicaSet`s, because it looks for `RunnerReplicaSet`s by labels but the existing `RunnerReplicaSet`s  are not labeled.
I guess that new `RunnerReplicaSet`s will be created after the controller is upgraded.

I found 2 ways to avoid that.

- The controller lists `RunnerReplicaSet`s both with `labels` and `ownerReference` and makes a set of both.
- The controller lists `RunnerReplicaSet`s first, and if they are not labeled, it labels `RunnerReplicaSet`s first.
  Then, it lists `RunnerReplicaSet`s both with `labels` again.

These are temporal. and you can delete them when you are forced to break backward compatibility.

Reply:
@tapih Thanks! Could you double-check my idea?

> The new controller cannot find the existing RunnerReplicaSets, because it looks for RunnerReplicaSets by labels but the existing RunnerReplicaSets are not labeled.
>I guess that new RunnerReplicaSets will be created after the controller is upgraded.

Would this be about the updated controller results in recreating [RunnerReplicaSet due to the replicasets' template hash change](https://github.com/summerwind/actions-runner-controller/blob/master/controllers/runnerdeployment_controller.go#L134) by the addition of the selector? As long as the controller update doesn't result in requiring the user to manually update K8s resources other than the controller's, it looks fine to me. WDYT?

My motivation behind https://github.com/summerwind/actions-runner-controller/commit/62fcd5f3de285f2c1384406fd0efbbfad3f01a92 was to keep backward-compatibility in the sense the sense of the RunnerDeployment API / avoid [forcing users to manually update their RunnerDeployment resources to include Spec.Template.Selector](https://github.com/summerwind/actions-runner-controller/commit/bbfb0b6bc137d077cf2ba159864bed148884ea5e#diff-4b55da611104d855c66d352a05727ec60b647333a87e0401e825bd5d1cb0e489R316).

Reply:
`desiredRS` would change due to the addition of the selector, but [these lines](https://github.com/tapih/actions-runner-controller/blob/8de84f9e3f62fbc90d6fd580f5fb9c1501cc99d9/controllers/runnerdeployment_controller.go#L82-L85) in the current code does not get the existing `RunnerReplicaSet`s in the first place because it now matches `RunnerReplicaSet`s with labels but `RunnerReplicaSet`s are not labeled before the controller is updated, IIUC.
I understand your motivation here and agree with you, so we should do what I suggested [here](https://github.com/summerwind/actions-runner-controller/pull/355#discussion_r583381120) to avoid this problem, .

What do you think about it? Thank you.

Reply:
@tapih Thanks for confirming!

What if we reverted that line to https://github.com/summerwind/actions-runner-controller/blob/master/controllers/runnerdeployment_controller.go#L79 so that we can keep using controller-runtime's index for searching RunnerReplicaSets owned by the RunnerDeployment?

That would still make the updated controller automatically replacing RunnerReplicaSets due to the selector changes, with neither additional K8s API call nor user intervention. WDYT?

Reply:
@mumoshu 

Yes, it seems to work correctly enough, keeping the backward compatibility.
I also considered what happens on `RunnerReplicaSet` and `Runner`, and thought that it also works correctly because
the old `RunnerReplicaSet`s are deleted [here](https://github.com/summerwind/actions-runner-controller/blob/master/controllers/runnerdeployment_controller.go#L178-L191) and the existing `Runner`s are deleted eventually by the controller, referencing `ownerReference`.

The Kubernetes native `Deployment` manages `ReplicaSet` with labels, so I still prefer to do the same thing for `RunnerDeployment` at a certain point, but your idea is reasonable for now with consideringg the compatibility

Reply:
@tapih Thanks again for the confirmation and the update 9f8220c!

I took some more time to think about this. Let me confirm our direction:

### HRA controller

The HRA controller should use the label selector for querying runners from HRA for autoscaling. For that, we'd also need to propagate labels from RunnerDeployment to Runners via RunnerReplicaSet. Currently, it lists the all runners in a namespace for calculation, which is wrong.

### RunnerDeployment controller

The RunnerDeployment controller should keep using the owner-reference index for querying replica sets. By doing that the controller can reliably delete the old RS and create new RS on controller update.

### RunnerReplicaSet controller

Similarly, RunnerReplicaSet controller should use the owner-reference index for querying runners. Currently, it lists all the runners in a namespace which is definitely wrong. This change allows the controller to keep tracking all the runners before/after the controller update, while extra support for multiple RunnerReplicaSet per namespace.

### Future work

The new RunnerReplicaSet and Runner will get more labels after the controller update.

In the future, the controller may use those labels for querying runners and pods. This should make our controller more like K8s deployment controller(ref https://github.com/summerwind/actions-runner-controller/pull/355#discussion_r584473993)

---

I believe that the label propagation and the RunnerDeployment controller changes are already done thanks to you. I'll update the RunnerReplicaSet controller to use the owner-reference index.

Reply:
desiredRS changes due to the addition of the selector, but apparently it doesn't trigger the replacement of the RS managed by the deployment, as it is triggered only when the runner template hash changes. The hash is calculated from RunnerDeployment.Spec.Template, not RunnerDeployment.Spec.Selector or the [default selector populated by the controller](https://github.com/summerwind/actions-runner-controller/blob/9f8220ced3b373d6fd64def36370494f24e05acf/controllers/runnerdeployment_controller.go#L316).

Probably this is the correct behaviour considering how the standard K8s deployment controller works. https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#updating-a-deployment

But doesn't it break https://github.com/summerwind/actions-runner-controller/blob/9f8220ced3b373d6fd64def36370494f24e05acf/controllers/runnerreplicaset_controller.go on the replicaset controller, as the selector is not populated on the existing replicaset after the controller update, and the repiicaset is not automatically replaced by the updated controller?

Reply:
Thank you for wrapping up!

One thing I would like to confirm is that you canceled the `Similarly...` line, and that means `RunnerReplicaSet` manages `Runner` with labels? If so, I'm gonna make another small fix for HRA to work when no selector is given.

Reply:
I'm not entirely sure yet but I think we need 9d75e30

Reply:
I'm not entirely sure yet but I think we need 9d75e30. Would you agree @tapih?

Reply:
The reason I canceled `Similarly...` line was that it was my total misconception. The replicaset controller was not using the owner-reference index but it was filtering runners by the owner reference on its own.

> If so, I'm gonna make another small fix for HRA to work when no selector is given.

Would you think that it's needed even after https://github.com/summerwind/actions-runner-controller/commit/9d75e30c9cd39b9f46aca5afd3b56caec1a99d27?

Reply:
> RunnerReplicaSet manages Runner with labels?

Thanks to your change, [RunnerReplicaSet already seems to manage Runner with labels](https://github.com/summerwind/actions-runner-controller/blob/9d75e30c9cd39b9f46aca5afd3b56caec1a99d27/controllers/runnerreplicaset_controller.go#L81). After 9d75e30, the controller even works correctly for RunnerReplicaSet with/without Spec.Selector.

> If so, I'm gonna make another small fix for HRA to work when no selector is given.

Are you willing to add something like the owner-reference check similar to https://github.com/summerwind/actions-runner-controller/commit/9d75e30c9cd39b9f46aca5afd3b56caec1a99d27 the below?

https://github.com/summerwind/actions-runner-controller/blob/9d75e30c9cd39b9f46aca5afd3b56caec1a99d27/controllers/runnerreplicaset_controller.go#L77-L81

Reply:
@tapih Just a friendly reminder for the case that you missed my comment 😃

Reply:
Sorry, I took a day off yesterday...

Reply:
I took this way to fix the HRA bug because I thought we should basically use labels to have `RunnerDeployment` manage `Runner`s to make the logic as symmetric as possible with k8s's `Deployment`
At the same time, I undestood that you didn't want to break the compatibility, so I made some patches not to break the compatibility afterwards.

However, now the code is getting more and more complex, and I'm being strongly concerned that this would make it unnecessaliry harder to maintain this project later.
I'm really sorry for taking your time so much, but I would like to suggest taking the following steps:

1. Fix the HRA bug without using labels
2. Merge the PR which has the `RunnerDeployment` manage `Runner`s with labels when you are forced to break the compatibility later

I'm willing to make PRs again. What do you think about it?

Reply:
@tapih Just to be clear, I was actually inclined to add `selector` to RunnerDeployment and RunnerReplicaSet, despite that it requires some work to make it backward-compatible.

Not only it helps us shape RunnerDeplomyet as much as similar as K8s Deployment, I found that `selector` itself can be a good/common language that allows the third-party(user) to easily find a specific set of runners managed by a runnerreplicaset/runnerdeployment.

Without `selector`, one had to read actions-runner-controller code to see what information can be used to correlate a runner deployment/replicaset to the corresponding runner(s).

So we now seem to have another benefit that we weren't aware of, for adding `selector`.

In addition to that-

> Fix the HRA bug without using labels

(I may have been pretty confused, but) if you literally meant to NOT use labels, I'm not sure how we can achieve that. If you meant to not use selectors, I'm still unsure how easy it is.

HRA needs to find the latest runners managed by the targeted RunnerDeployment anyway.

Just following owner-references from HRA, RD, RS, and Runners would be not enough.

For HRA-RD, RS-R it's fine, as there might be only 1-to-1 relationship among them. However for RD, there may be two or more RS. They have different `runner-template-hash` label values and that label has been already there for RunnerReplicaSet, so using it for filtering RunnerReplicaSet seems natural.

We also need to determine which `runner-template-hash` value is one for the latest runner spec. Instead of relying on `Selector`, we can probably modify the HRA controller to call the same hash computing function on the RD.Spec.Template as the RD controller for that. But doing so would make a tight coupling between RD and HRA controllers. Wouldn't it be easier to add/use `Selector` then, in terms of design/implementation complexity?

WDYT?

You may have thought if it's really up to me to decide which direction we'd go, but I'd really appreciate your feedback!